### PR TITLE
[111] Build configurations and custom flags

### PR DESCRIPTION
**Pull request by phongvhd93**
https://github.com/nimblehq/ios-templates/issues/111

## What happened

As an iOS developer, I can see 6 build configurations and 4 custom flags after create project from iOS project templates.

**Flags and build configurations must follow following rules:**

Configurations / Custom flags | DEBUG | UAT | Staging | Production
-- | -- | -- | -- | --
Dev UAT | x | x |   |  
UAT |   | x |   |  
Dev Staging | x |   | x |  
Staging |   |   | x |  
Dev Production | x |   |   | x
Production |   |   |   | x

## Insight

Added build configurations:
 - Dev Staging
 - Staging
 - Dev UAT
 - UAT
 - Dev Production
 - Production

Add flags:
 - DEBUG
 - UAT
 - STAGING
 - PRODUCTION
 
## Proof Of Work

Screen Shot 2020-10-15 at 16 03 08

### CODE REVIEW

File: Nimble Templates/App.xctemplate/TemplateInfo.plist [hunk 1]
```diff
@@ -1,215 +1,391 @@
...
-	</array>
-	<key>Concrete</key>
-	<true/>
-	<key>Description</key>
-	<string>This template provides a starting point for an application that uses a single view.</string>
-	<key>SortOrder</key>
-	<integer>1</integer>
-	<key>Options</key>
-	<array>
+	<dict>
```
Comment (suho) [inline Nimble Templates/App.xctemplate/TemplateInfo.plist:4]: Please check the indent here

File: Nimble Templates/App.xctemplate/TemplateInfo.plist [hunk 2]
```diff
@@ -1,215 +1,391 @@
...
+		<array>
+			<dict>
+				<key>Name</key>
+				<string>___PACKAGENAME___</string>
+				<key>SharedSettings</key>
+				<dict>
+					<key>PRODUCT_NAME</key>
+					<string>$(TARGET_NAME)</string>
+					<key>SWIFT_VERSION</key>
+					<string>5.0</string>
```
Comment (suho) [inline Nimble Templates/App.xctemplate/TemplateInfo.plist:177]: Please check the version here, when using the Xcode 12 to create a new project, what is the version here?

File: Nimble Templates/App.xctemplate/TemplateInfo.plist [hunk 3]
```diff
@@ -1,215 +1,391 @@
...
 					<dict>
-						<key>RequiredOptions</key>
-						<dict>
-							<key>userInterface</key>
-							<string>Programmatically</string>
-						</dict>
+						<key>Class</key>
+						<string>ShellScript</string>
+						<key>Name</key>
+						<string>Run SwiftLint</string>
```
Comment (suho) [inline Nimble Templates/App.xctemplate/TemplateInfo.plist:185]: Please remove SwiftLint in this pull request

File: Nimble Templates/App.xctemplate/TemplateInfo.plist [hunk 4]
```diff
@@ -1,215 +1,391 @@
...
 		<dict>
-			<key>Default</key>
-			<string>Swift</string>
-			<key>Identifier</key>
-			<string>languageChoice</string>
-			<key>Units</key>
+			<key>Configurations</key>
 			<dict>
-				<key>Objective-C</key>
+				<key>Dev Staging</key>
```
Comment (suho) [inline Nimble Templates/App.xctemplate/TemplateInfo.plist:24]: Please update the README.md if we need to close the project and open it again to see the change

**Review by suho — COMMENTED** [review]
Please check my comment, thank you.

File: Nimble Templates/App.xctemplate/TemplateInfo.plist (hunk 2 above)
Comment (phongvhd93) [reply Nimble Templates/App.xctemplate/TemplateInfo.plist]: It is 5.0 on Xcode 12

**Review by suho — APPROVED** [review]
LGTM

**Review by blyscuit — COMMENTED** [review]
I think we should also remove Build Settings other than flags from this commit.

**Comment by phongvhd93** [comment]
> I think we should also remove Build Settings other than flags from this commit.

Could you provide me some more context, I don't get your ideas.

File: Nimble Templates/App.xctemplate/TemplateInfo.plist [hunk 5]
```diff
@@ -16,6 +16,167 @@
...
+				<string>NO</string>
+				<key>GCC_OPTIMIZATION_LEVEL</key>
+				<string>0</string>
+				<key>SWIFT_COMPILATION_MODE</key>
+				<string>wholemodule</string>
+				<key>SWIFT_OPTIMIZATION_LEVEL</key>
+				<string>-Onone</string>
+				<key>DEBUG_INFORMATION_FORMAT</key>
+				<string>dwarf-with-dsym</string>
+				<key>SWIFT_ACTIVE_COMPILATION_CONDITIONS</key>
```
Comment (blyscuit) [inline Nimble Templates/App.xctemplate/TemplateInfo.plist:65]: This area includes many unnecessary `Build Settings` config, we could remove them as they are not use for this template. You can see from `line number 24` and decide on what we want to keep. They are repeated in most of the `Configurations` 👍

File: Nimble Templates/App.xctemplate/TemplateInfo.plist [hunk 6]
```diff
@@ -16,6 +16,167 @@
...
+				<key>SWIFT_OPTIMIZATION_LEVEL</key>
+				<string>-Onone</string>
+				<key>SWIFT_ACTIVE_COMPILATION_CONDITIONS</key>
+				<array>
+					<string>DEBUG</string>
+					<string>STAGING</string>
+				</array>
+				<key>DEBUG_INFORMATION_FORMAT</key>
+				<string>dwarf</string>
+				<key>PRODUCT_BUNDLE_IDENTIFIER</key>
```
Comment (nmint8m) [inline Nimble Templates/App.xctemplate/TemplateInfo.plist:46]: The attribute `PRODUCT_BUNDLE_IDENTIFIER` does not work, it create duplicate values, please check again.

**Comment by nmint8m** [comment]
Some Xcode project's constants that we need to know: https://www.rubydoc.info/gems/xcodeproj/Xcodeproj/Constants

When the template inherits from some ancestors, we may face unexpected build setting attributes which are automatically initialized by the ancestors.

For example:

```
PROJECT_DEFAULT_BUILD_SETTINGS =
Returns The default build settings for a new project.

Returns:

(Hash) — The default build settings for a new project.
{
  :all => {
    'ALWAYS_SEARCH_USER_PATHS'                => 'NO',
    'CLANG_ANALYZER_NONNULL'                  => 'YES',
    'CLANG_ANALYZER_NUMBER_OBJECT_CONVERSION' => 'YES_AGGRESSIVE',
    'CLANG_CXX_LANGUAGE_STANDARD'             => 'gnu++14',
    'CLANG_CXX_LIBRARY'                       => 'libc++',
    'CLANG_ENABLE_MODULES'                    => 'YES',
    'CLANG_ENABLE_OBJC_ARC'                   => 'YES',
    'CLANG_ENABLE_OBJC_WEAK'                  => 'YES',
    'CLANG_WARN__DUPLICATE_METHOD_MATCH'      => 'YES',
    'CLANG_WARN_BLOCK_CAPTURE_AUTORELEASING'  => 'YES',
    'CLANG_WARN_BOOL_CONVERSION'              => 'YES',
    'CLANG_WARN_COMMA'                        => 'YES',
    'CLANG_WARN_CONSTANT_CONVERSION'          => 'YES',
    'CLANG_WARN_DEPRECATED_OBJC_IMPLEMENTATIONS' => 'YES',
    'CLANG_WARN_DIRECT_OBJC_ISA_USAGE'        => 'YES_ERROR',
    'CLANG_WARN_DOCUMENTATION_COMMENTS'       => 'YES',
    'CLANG_WARN_EMPTY_BODY'                   => 'YES',
    'CLANG_WARN_ENUM_CONVERSION'              => 'YES',
    'CLANG_WARN_INFINITE_RECURSION'           => 'YES',
    'CLANG_WARN_INT_CONVERSION'               => 'YES',
    'CLANG_WARN_NON_LITERAL_NULL_CONVERSION'  => 'YES',
    'CLANG_WARN_OBJC_IMPLICIT_RETAIN_SELF'    => 'YES',
    'CLANG_WARN_OBJC_LITERAL_CONVERSION'      => 'YES',
    'CLANG_WARN_OBJC_ROOT_CLASS'              => 'YES_ERROR',
    'CLANG_WARN_QUOTED_INCLUDE_IN_FRAMEWORK_HEADER' => 'YES',
    'CLANG_WARN_RANGE_LOOP_ANALYSIS'          => 'YES',
    'CLANG_WARN_STRICT_PROTOTYPES'            => 'YES',
    'CLANG_WARN_SUSPICIOUS_MOVE'              => 'YES',
    'CLANG_WARN_UNGUARDED_AVAILABILITY'       => 'YES_AGGRESSIVE',
    'CLANG_WARN_UNREACHABLE_CODE'             => 'YES',
    'COPY_PHASE_STRIP'                        => 'NO',
    'ENABLE_STRICT_OBJC_MSGSEND'              => 'YES',
    'GCC_C_LANGUAGE_STANDARD'                 => 'gnu11',
    'GCC_NO_COMMON_BLOCKS'                    => 'YES',
    'GCC_WARN_64_TO_32_BIT_CONVERSION'        => 'YES',
    'GCC_WARN_ABOUT_RETURN_TYPE'              => 'YES_ERROR',
    'GCC_WARN_UNDECLARED_SELECTOR'            => 'YES',
    'GCC_WARN_UNINITIALIZED_AUTOS'            => 'YES_AGGRESSIVE',
    'GCC_WARN_UNUSED_FUNCTION'                => 'YES',
    'GCC_WARN_UNUSED_VARIABLE'                => 'YES',
    'MTL_FAST_MATH'                           => 'YES',
    'PRODUCT_NAME'                            => '$(TARGET_NAME)',
    'SWIFT_VERSION'                           => '5.0',
  },
  :release => {
    'DEBUG_INFORMATION_FORMAT'           => 'dwarf-with-dsym',
    'ENABLE_NS_ASSERTIONS'               => 'NO',
    'MTL_ENABLE_DEBUG_INFO'              => 'NO',
    'SWIFT_COMPILATION_MODE'             => 'wholemodule',
    'SWIFT_OPTIMIZATION_LEVEL'           => '-O',
  }.freeze,
  :debug => {
    'DEBUG_INFORMATION_FORMAT'            => 'dwarf',
    'ENABLE_TESTABILITY'                  => 'YES',
    'GCC_DYNAMIC_NO_PIC'                  => 'NO',
    'GCC_OPTIMIZATION_LEVEL'              => '0',
    'GCC_PREPROCESSOR_DEFINITIONS'        => ['DEBUG=1', '$(inherited)'],
    'MTL_ENABLE_DEBUG_INFO'               => 'INCLUDE_SOURCE',
    'ONLY_ACTIVE_ARCH'                    => 'YES',
    'SWIFT_ACTIVE_COMPILATION_CONDITIONS' => 'DEBUG',
    'SWIFT_OPTIMIZATION_LEVEL'            => '-Onone',
  }.freeze,
}.freeze
```

So we should check:
- Which one contain those constants.
- How to override them.

**Comment by nmint8m** [comment]
One more thing, since we have no solution for (this comment)[https://github.com/nimblehq/ios-templates/pull/117#discussion_r507883871], so I will write a note to remove duplicate build settings manually within my PR. 👍

**Review by nmint8m — APPROVED** [review]

**Comment by phongvhd93** [comment]
> Some Xcode project's constants that we need to know: https://www.rubydoc.info/gems/xcodeproj/Xcodeproj/Constants
> 
> When the template inherits from some ancestors, we may face unexpected build setting attributes which are automatically initialized by the ancestors.
> 
> For example:
> 
> ```
> PROJECT_DEFAULT_BUILD_SETTINGS =
> Returns The default build settings for a new project.
> 
> Returns:
> 
> (Hash) — The default build settings for a new project.
> {
>   :all => {
>     'ALWAYS_SEARCH_USER_PATHS'                => 'NO',
>     'CLANG_ANALYZER_NONNULL'                  => 'YES',
>     'CLANG_ANALYZER_NUMBER_OBJECT_CONVERSION' => 'YES_AGGRESSIVE',
>     'CLANG_CXX_LANGUAGE_STANDARD'             => 'gnu++14',
>     'CLANG_CXX_LIBRARY'                       => 'libc++',
>     'CLANG_ENABLE_MODULES'                    => 'YES',
>     'CLANG_ENABLE_OBJC_ARC'                   => 'YES',
>     'CLANG_ENABLE_OBJC_WEAK'                  => 'YES',
>     'CLANG_WARN__DUPLICATE_METHOD_MATCH'      => 'YES',
>     'CLANG_WARN_BLOCK_CAPTURE_AUTORELEASING'  => 'YES',
>     'CLANG_WARN_BOOL_CONVERSION'              => 'YES',
>     'CLANG_WARN_COMMA'                        => 'YES',
>     'CLANG_WARN_CONSTANT_CONVERSION'          => 'YES',
>     'CLANG_WARN_DEPRECATED_OBJC_IMPLEMENTATIONS' => 'YES',
>     'CLANG_WARN_DIRECT_OBJC_ISA_USAGE'        => 'YES_ERROR',
>     'CLANG_WARN_DOCUMENTATION_COMMENTS'       => 'YES',
>     'CLANG_WARN_EMPTY_BODY'                   => 'YES',
>     'CLANG_WARN_ENUM_CONVERSION'              => 'YES',
>     'CLANG_WARN_INFINITE_RECURSION'           => 'YES',
>     'CLANG_WARN_INT_CONVERSION'               => 'YES',
>     'CLANG_WARN_NON_LITERAL_NULL_CONVERSION'  => 'YES',
>     'CLANG_WARN_OBJC_IMPLICIT_RETAIN_SELF'    => 'YES',
>     'CLANG_WARN_OBJC_LITERAL_CONVERSION'      => 'YES',
>     'CLANG_WARN_OBJC_ROOT_CLASS'              => 'YES_ERROR',
>     'CLANG_WARN_QUOTED_INCLUDE_IN_FRAMEWORK_HEADER' => 'YES',
>     'CLANG_WARN_RANGE_LOOP_ANALYSIS'          => 'YES',
>     'CLANG_WARN_STRICT_PROTOTYPES'            => 'YES',
>     'CLANG_WARN_SUSPICIOUS_MOVE'              => 'YES',
>     'CLANG_WARN_UNGUARDED_AVAILABILITY'       => 'YES_AGGRESSIVE',
>     'CLANG_WARN_UNREACHABLE_CODE'             => 'YES',
>     'COPY_PHASE_STRIP'                        => 'NO',
>     'ENABLE_STRICT_OBJC_MSGSEND'              => 'YES',
>     'GCC_C_LANGUAGE_STANDARD'                 => 'gnu11',
>     'GCC_NO_COMMON_BLOCKS'                    => 'YES',
>     'GCC_WARN_64_TO_32_BIT_CONVERSION'        => 'YES',
>     'GCC_WARN_ABOUT_RETURN_TYPE'              => 'YES_ERROR',
>     'GCC_WARN_UNDECLARED_SELECTOR'            => 'YES',
>     'GCC_WARN_UNINITIALIZED_AUTOS'            => 'YES_AGGRESSIVE',
>     'GCC_WARN_UNUSED_FUNCTION'                => 'YES',
>     'GCC_WARN_UNUSED_VARIABLE'                => 'YES',
>     'MTL_FAST_MATH'                           => 'YES',
>     'PRODUCT_NAME'                            => '$(TARGET_NAME)',
>     'SWIFT_VERSION'                           => '5.0',
>   },
>   :release => {
>     'DEBUG_INFORMATION_FORMAT'           => 'dwarf-with-dsym',
>     'ENABLE_NS_ASSERTIONS'               => 'NO',
>     'MTL_ENABLE_DEBUG_INFO'              => 'NO',
>     'SWIFT_COMPILATION_MODE'             => 'wholemodule',
>     'SWIFT_OPTIMIZATION_LEVEL'           => '-O',
>   }.freeze,
>   :debug => {
>     'DEBUG_INFORMATION_FORMAT'            => 'dwarf',
>     'ENABLE_TESTABILITY'                  => 'YES',
>     'GCC_DYNAMIC_NO_PIC'                  => 'NO',
>     'GCC_OPTIMIZATION_LEVEL'              => '0',
>     'GCC_PREPROCESSOR_DEFINITIONS'        => ['DEBUG=1', '$(inherited)'],
>     'MTL_ENABLE_DEBUG_INFO'               => 'INCLUDE_SOURCE',
>     'ONLY_ACTIVE_ARCH'                    => 'YES',
>     'SWIFT_ACTIVE_COMPILATION_CONDITIONS' => 'DEBUG',
>     'SWIFT_OPTIMIZATION_LEVEL'            => '-Onone',
>   }.freeze,
> }.freeze
> ```
> 
> So we should check:
> 
> * Which one contain those constants.
> * How to override them.

This came from `Ancestor` of template, I think we need to take time to find out which one made this. It seems we must search a lot in Package content of XCode.

**Review by blyscuit — APPROVED** [review]

**Pull request merged**